// @fabcotech/rchain-toolkit/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.publicKeyFromPrivateKey = exports.getDeployOptions = exports.transferRevTerm = exports.signSecp256k1 = exports.verifyPrivateAndPublicKey = exports.toByteArray = exports.varToRhoExpr = exports.mapToRhoRepr = exports.unforgeableWithId = exports.getValueFromBlocks = exports.getFirstBlock = void 0;
const elliptic = __importStar(require("elliptic"));
const rnodeProtos = __importStar(require("./rnode-protos"));
const getDeployData_1 = require("./utils/getDeployData");
const getDeployDataToSign_1 = require("./utils/getDeployDataToSign");
const getBlake2Hash_1 = require("./utils/getBlake2Hash");
__exportStar(require("./utils/rhoValToJs"), exports);
__exportStar(require("./utils/rhoExprToVar"), exports);
__exportStar(require("./utils/decodePar"), exports);
__exportStar(require("./utils/getBlake2Hash"), exports);
__exportStar(require("./utils/getDeployData"), exports);
__exportStar(require("./utils/getDeployDataToSign"), exports);
__exportStar(require("./utils/revAddressFromPublicKey"), exports);
__exportStar(require("./utils/ethAddressFromPublicKey"), exports);
const getFirstBlock = (blockInfo) => {
    return blockInfo[0];
};
exports.getFirstBlock = getFirstBlock;
const getValueFromBlocks = (blockInfo) => {
    for (let i = 0; i < blockInfo.length; i += 1) {
        const block = blockInfo[i];
        if (block.postBlockData) {
            for (let j = 0; j < block.postBlockData.length; j += 1) {
                const data = block.postBlockData[j];
                if (data) {
                    return data;
                }
            }
        }
    }
    throw new Error("Not data found in any block");
};
exports.getValueFromBlocks = getValueFromBlocks;
const unforgeableWithId = (id) => {
    return id.toString("hex");
};
exports.unforgeableWithId = unforgeableWithId;
const stringToRhoRepr = (a) => {
    return { g_string: a };
};
const intToRhoRepr = (a) => {
    return { g_int: a };
};
const boolToRhoRepr = (a) => {
    return { g_bool: a };
};
const listToRhoRepr = (a) => {
    return {
        ps: a.map((e) => ({ exprs: [(0, exports.varToRhoExpr)(e)] })),
    };
};
const mapToRhoRepr = (a) => {
    const map = {
        kvs: [],
    };
    Object.keys(a)
        .sort() // alphabetical
        .forEach((key) => {
        if (typeof a[key] !== "undefined" && a[key] !== null) {
            map.kvs.push({
                key: { exprs: [(0, exports.varToRhoExpr)(key)] },
                value: { exprs: [(0, exports.varToRhoExpr)(a[key])] },
            });
        }
        else {
            map.kvs.push({
                key: { exprs: [(0, exports.varToRhoExpr)(key)] },
            });
        }
    });
    return map;
};
exports.mapToRhoRepr = mapToRhoRepr;
const varToRhoExpr = (a) => {
    if (typeof a === "boolean") {
        return boolToRhoRepr(a);
    }
    if (typeof a === "string") {
        return stringToRhoRepr(a);
    }
    if (typeof a === "number") {
        return intToRhoRepr(a);
    }
    if (Array.isArray(a)) {
        return { e_list_body: listToRhoRepr(a) };
    }
    if (typeof a === "object" && a !== null) {
        return { e_map_body: (0, exports.mapToRhoRepr)(a) };
    }
    throw new Error("Unknown variable type");
};
exports.varToRhoExpr = varToRhoExpr;
const toByteArray = (a) => {
    const expr = (0, exports.varToRhoExpr)(a);
    return rnodeProtos.Par.encode({
        exprs: [expr],
    }).finish();
};
exports.toByteArray = toByteArray;
const verifyPrivateAndPublicKey = (privateKey, publicKey) => {
    const ec = new elliptic.ec("secp256k1");
    const keyPair = ec.keyFromPrivate(privateKey);
    if (keyPair.getPublic().encode("hex", false) !== publicKey) {
        throw new Error("Private key and public key do not match");
    }
};
exports.verifyPrivateAndPublicKey = verifyPrivateAndPublicKey;
const signSecp256k1 = (hash, privateKey) => {
    const ec = new elliptic.ec("secp256k1");
    const keyPair = ec.keyFromPrivate(privateKey);
    const signature = keyPair.sign((Buffer || Buffer).from(hash), {
        canonical: true,
    });
    const derSign = signature.toDER();
    if (!ec.verify(Buffer.from(hash), signature, keyPair, "hex")) {
        throw new Error("Failed to verify signature");
    }
    return new Uint8Array(derSign);
};
exports.signSecp256k1 = signSecp256k1;
const transferRevTerm = (a) => {
    if (typeof a.from !== "string") {
        throw new Error("from must be a REV address");
    }
    if (typeof a.to !== "string") {
        throw new Error("to must be a REV address");
    }
    if (typeof a.amount !== "number") {
        throw new Error("amount must be a number");
    }
    return `new
  deployId(\`rho:rchain:deployId\`),
  rl(\`rho:registry:lookup\`),
  RevVaultCh,
  stdout(\`rho:io:stdout\`)
in {

rl!(\`rho:rchain:revVault\`, *RevVaultCh) |
for (@(_, RevVault) <- RevVaultCh) {

  match (
    "${a.from}",
    "${a.to}",
    ${a.amount}
  ) {
    (from, to, amount) => {

      new vaultCh, revVaultkeyCh, deployerId(\`rho:rchain:deployerId\`) in {
        @RevVault!("findOrCreate", from, *vaultCh) |
        @RevVault!("deployerAuthKey", *deployerId, *revVaultkeyCh) |
        for (@(true, vault) <- vaultCh; key <- revVaultkeyCh) {

          stdout!(("Beginning transfer of ", amount, "REV from", from, "to", to)) |

          new resultCh in {
            @vault!("transfer", to, amount, *key, *resultCh) |
            for (@result <- resultCh) {
              stdout!(("Finished transfer of ", amount, "REV to", to, "result was:", result))
            }
          }
        }
      }
    }
  }
}
}`;
};
exports.transferRevTerm = transferRevTerm;
const getDeployOptions = (payload) => {
    const deployData = (0, getDeployData_1.getDeployData)({
        timestamp: payload.timestamp,
        shardId: payload.shardId || "",
        term: payload.term,
        phloPrice: typeof payload.phloPrice === 'number' ? payload.phloPrice : 1,
        phloLimit: typeof payload.phloLimit === 'number' ? payload.phloLimit : 10000,
        validAfterBlockNumber: typeof payload.validAfterBlockNumber === 'number' ? payload.validAfterBlockNumber : -1
    });
    const publicKey = (0, exports.publicKeyFromPrivateKey)(payload.privateKey);
    const toSign = (0, getDeployDataToSign_1.getDeployDataToSign)(deployData);
    const hash = (0, getBlake2Hash_1.getBlake2Hash)(toSign);
    let signature;
    signature = (0, exports.signSecp256k1)(hash, payload.privateKey);
    return {
        data: deployData,
        deployer: publicKey,
        signature: Buffer.from(new Uint8Array(signature)).toString("hex"),
        sigAlgorithm: payload.sigAlgorithm || "secp256k1",
    };
};
exports.getDeployOptions = getDeployOptions;
// Address and public key
// Algorithm to generate ETH and REV address is taken from RNode source
// https://github.com/rchain/rchain/blob/bf7a30e1d388d46aa9e5f4b8c04089fc8e31d771/rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala#L47
const publicKeyFromPrivateKey = (privateKey) => {
    const ec = new elliptic.ec("secp256k1");
    const keyPair = ec.keyFromPrivate(privateKey);
    return keyPair.getPublic().encode("hex", false);
};
exports.publicKeyFromPrivateKey = publicKeyFromPrivateKey;
